### ISSUE TITLE
Investigate flushing Onyx cache to disk on idle

### ISSUE THREAD

**Issue by marcaaron** [Expensify/react-native-onyx#281]
We have observed that IndexedDB can be super slow with writes and cause major lag and jank in the app when there is a huge amount of data to store on App init. 

Our solution for now has been to just completely cease storing certain keys entirely. But, another potentially viable solution is to keep the data in memory for some time and then persist it to disk when the user is "idling".

cc @hannojg @kidroca curious if you guys have considered this one yet.

h/t @quinthar for the idea

**Comment by kidroca** [comment]
Hi @marcaaron, 

Technically, it's not so much that `IndexedDB` is inherently slow, but more so that it's implemented inefficiently by `localforage` when subjected to frequent write operations.

To illustrate, as noted in this thread: https://expensify.slack.com/archives/C01GTK53T8Q/p1689267792230089?thread_ts=1689036681.307159&cid=C01GTK53T8Q:

> "Reviewing the `localForage` [source code for setItem](https://github.com/localForage/localForage/blob/c5b66157252988dc4e74a3eab5226b59be376a2f/src/drivers/indexeddb.js#L640C49-L640C49), we can observe that a new transaction is indeed opened each time the function is invoked. This could be a significant contributing factor to the noticeable slowdown experienced when executing `multiSet` or `multiMerge` operations. Currently, both these operations use `setItem` to process each item individually, which could potentially lead to high resource contention and decreased performance."

In light of this, I propose that we consider implementing a storage provider that circumvents the use of `localforage`. Instead, this provider would interface directly with the IndexedDB API, enabling it to more effectively strategize and optimize transactions. For instance, it could perform a single transaction to update multiple keys as opposed to sequential transactions. This would decrease the likelihood of lock contention and should, in theory, enhance speed significantly.
_(The change should also let us drop the `SyncQueue` as we should be able to write in one go, not having to schedule writes one by one)_

Alternatively, we could explore other storage driver libraries that might be a better fit for our needs. There exist libraries that streamline the IndexedDB interface, offering ease of use without forfeiting granular control.

I like the idea of batching a few changes and applying them at the best time. It appears that the app does not require immediate data writes. In the worst-case scenario, we might "skip of a beat," and some information might need to be fetched remotely due to delayed persistence.

**Comment by marcaaron** [comment]
Great thoughts @kidroca. So it seems like our options are:

- Investigate writing our own IndexedDB implementation (Feels potentially high complexity, but I lack experience working with the API directly so not sure)
- Investigate an existing API wrapper around IndexedDB besides `localforage` (Does it exist?)
- Keep using `localforage`, but defer writes until idle

I do like this 3rd option since it does not seem to difficult to implement something like:

```js
let timer;
const idleTime = 2000; // or whatever we want to use for this

function resetTimer() {
  pauseSetItemQueue();
  clearTimeout(timer);
  timer = setTimeout(processSetItemQueue, idleTime);
}

document.addEventListener('mousemove', resetTimer);
document.addEventListener('keypress', resetTimer);
```

**Comment by kidroca** [comment]
@marcaaron I think this library might work better for our case: https://www.npmjs.com/package/idb-keyval the following methods are noteworthy:
- `setMany`: Set many keyval pairs at once. This is faster than calling set multiple times.
- `update`: gives access to the prev value similar to how `setState` works in react 

Regarding deferring until idle, there's this API available to browsers as well as react-native: https://caniuse.com/requestidlecallback - `requestIdleCallback` 

It accepts callback function that will be executed when the browser becomes idle 
There's also a numeric timeout parameter in case we want to set a condition like _"execute when idle, but no later than `{n}` milliseconds from now"_
Requests are otherwise served in a first in first out basis

Safari lacks support for `requestIdleCallback`, but there's a polyfill to cover the missing API

**Comment by marcaaron** [comment]
`requestIdleCallback` looks pretty neat.

**Comment by hannojg** [comment]
Hey, I somehow overlooked this ticket and was doing a performance audit on a guides account and came to the exact same issue of the massive amount of data we are persisting at app init.
Here are my thoughts so far (some of them are maybe out of the box, watch out 🧙‍♂️):

1. As @kidroca says, `indexedDB` itself can be pretty fast. The approach I think would be the most valuable to explore is to use indexedDB directly (or a neater layer if that really exists), and optimise the usage for how we need it. I believe this can bring the most performance improvements (and I have a strong itch to start writing a PoC 😄). Why do I believe it can bring the most performance improvements? I just measured it and for my guide account there are over 100.000 (yes) calls to `storage.setItem`. This also completely spams the microtask queue.
Batching all of that together (or at least in a few big transactions) should get most of the performance back I believe.
2. Move onyx to another thread using web workers. See [my thought process about it here](https://expensify.slack.com/archives/C035J5C9FAP/p1690816532748079). The main idea is that it won't matter to congest the micro task queue if it happens on a different thread (however, this has the downside of the persistence still taking super long and we might loose data if the user closes the browser window prematurely.)
3. When it comes to running task only when the browser gets idle I think there are better solutions. Check out this code example, which utilises [`isInputPending()`](https://developer.chrome.com/articles/isinputpending/)
```js
async function saveSettings () {
  // A task queue of functions
  // Imagine this being tasks e.g. inside onyx where we we execute database operations.
  const tasks = [
    validateForm,
    showSpinner,
    saveToDatabase,
    updateUI,
    sendAnalytics
  ];
  
  // The value of 50 is important here.
  // Any task that takes longer than 50ms is considered
  // too long and blocking. So we don't want our tasks to
  // take longer than that (so the browser has a chance to
  // squeeze in other tasks e.g. handling user interactions)
  let deadline = performance.now() + 50;

  while (tasks.length > 0) {
    // Optional chaining operator used here helps to avoid
    // errors in browsers that don't support `isInputPending`:
    if (navigator.scheduling?.isInputPending() || performance.now() >= deadline) {
      // There's a pending user input, or the
      // deadline has been reached. Yield here:
      await yieldToMain();

      // Extend the deadline:
      deadline = performance.now() + 50;

      // Stop the execution of the current loop and
      // move onto the next iteration:
      continue;
    }

    // Shift the task out of the queue:
    const task = tasks.shift();

    // Run the task:
    task();
  }
}```

**Comment by marcaaron** [comment]
>Move onyx to another thread using web workers

Curious to hear more about this. So, would we just defer writes to the worker thread by posting data to the worker and then have those messages get written to IndexedDB? Does it work the same when reading data?

**Comment by marcaaron** [comment]
>this has the downside of the persistence still taking super long and we might loose data if the user closes the browser window prematurely

I am not sure how to get around that. Either have to save the data faster or expect to lose some of it 😄 

Losing it might not be much of a concern depending on what is lost, when it would be refreshed again, and what bad things could happen if the data is not persisted (maybe nothing).

**Comment by marcaaron** [comment]
Maybe as a next step we can try to test idb-keyval with `setMany()` on one of these inundated accounts to gauge real-world web improvements? Let's also test mobile web? 

That would probably give us an idea of what kind of performance improvements we might see from getting off localforage (which I think is kind of a separate issue). 

Then we can explore throttling storage after?

FWIW, I think web workers are pretty cool, but most hesitant about this idea since it will add some build complexity.

**Comment by hannojg** [comment]
Hey 👋  I'd like to take the initiative to test with idb-keyval as we are currently hired to improve the performance, and this would just be my next step 😊 

Also I'd still would like to test a OMT performance PoC.

 > Then we can explore throttling storage after?

Yes! (You are right this is a bit of a separate issue we are talking about atm, but I think that should come first)

> So, would we just defer writes to the worker thread by posting data to the worker and then have those messages get written to IndexedDB? Does it work the same when reading data?

Yes, the worker thread would be concerned with writing the data. I also thought an about it reading the data. It might adds a tiny little overhead but then we are keeping more work away from the main thread which is the whole idea of the change

**Comment by hannojg** [comment]
Results for 1.000 items:

Onyx right now: 6500ms
Onyx with new lib: 143ms

**Comment by marcaaron** [comment]
Wow. What are we waiting for??? 😄

Seems clearly much better. Next thought would be...

Can we use the same DB? Or do we need some migration to move from `localforage` to `idb-keyval`?

Looks very promising though so I think we should go for it.

**Comment by hannojg** [comment]
I think no migration is needed. I will write a PR today, and check the results for when used in actual product 😊

**Comment by hannojg** [comment]
PR is up 😊 

- https://github.com/Expensify/react-native-onyx/pull/293

On another note in regards to the "Flush onyx cache to disk on idle".
I think another strategy to go about this is to just ignore waiting for the writes to finish. Right now we are using the sequential queue, and the queue gets blocked by waiting for an action to complete its write to the database.
The important thing to note here is, that writing to the database isn't a heavy operation. On both native and web we just pass our data to the database provider and from there on it's handled in "the background".
However, we are awaiting the finishing of the promises of these write operations, and block the rest of our UI (onyx) operations.
I believe we just shouldn't do that. Internally we can have a queue that makes sure data writes get executed in order, however, we shouldn't block the app.

To give you an example, even with `idb-keyval` writing the data of OpenApp to Onyx takes ~3 seconds. In these three seconds, no other network request (such as OpenReport) will continue executing as we are waiting for the task to finish.

(However! In the advent of idb-keyval, where operations usually take ~10ms, and pagination reducing the size of the network responses, I am not sure how big the benefit of these changes would actually be 🤔 )

**Comment by marcaaron** [comment]
>To give you an example, even with idb-keyval writing the data of OpenApp to Onyx takes ~3 seconds. In these three seconds, no other network request (such as OpenReport) will continue executing as we are waiting for the task to finish.

Hmm if this is true that I agree and not sure I see a clear reason why we would do this. The storage should be able to be "fire and forget". Nothing needs to "wait" on it. FWIW I don't think we were always doing this and if we started at some point it might not have been intentional.

I think we are talking about this code here? 

https://github.com/Expensify/App/blob/264d24939c8eff0de99cc81d084742c91ba84ef3/src/libs/Middleware/SaveResponseInOnyx.js#L17-L37

If these "update promises" would resolve after the data has went into the cache then we should not be concerned with whether they are written to storage.

**Comment by hannojg** [comment]
Yes, I am talking about that code. because then we have the network's sequential queue here:

https://github.com/Expensify/App/blob/ff423aa61fde128e981ee071b01011238927516c/src/libs/Network/SequentialQueue.js#L40-L44

which basically waits until the request is done including to process all middlewares, which includes writing to the disk. Only once that's done we call `process()` again to process the next request.

**Comment by marcaaron** [comment]
I don't have a great understanding of _why_ we are waiting there. I _think_ we were not always waiting and the change was introduced in [this PR](https://github.com/Expensify/App/pull/16545/files).

Gonna tag in @neil-marcellini and @jasperhuangg to see if they are able to help.

**Comment by marcaaron** [comment]
>Correct me if I'm wrong, but I feel like once a request is queued its params can't change, even if those params depend on Onyx values that were changed by a a previous request that wrote to Onyx.

That is correct. But would you mind clarifying how it relates to this convo? I am having trouble connecting the dots. Thanks!

>Is there any reason why we would forgo using the new library and remove this "wait for write" stuff instead?

They're loosely related. We're really discussing two separate considerations in this issue thread.

**Comment by neil-marcellini** [comment]
> we have the network's sequential queue here:
> 
> https://github.com/Expensify/App/blob/ff423aa61fde128e981ee071b01011238927516c/src/libs/Network/SequentialQueue.js#L40-L44
> 
> which basically waits until the request is done including to process all middlewares, which includes writing to the disk. Only once that's done we call `process()` again to process the next request.

Yeah that's been in place for a while, and it was actually introduce in this PR https://github.com/Expensify/App/pull/9306. As explained in the comment I implemented that bit of code so that getCurrentRequest works.

I did add also add the [queueOnyxUpdates](https://github.com/Expensify/App/blob/27b12b9da7bcb5a7ce67f3a03dd9f148dc81e62a/src/libs/actions/QueuedOnyxUpdates.js#L16-L18) updateHandler. I don't know if we could safely continue with the next request before the data is written to disk, but I'm guessing we probably could.

**Comment by jasperhuangg** [comment]
@marcaaron I was responding to this secondary consideration:

> I think another strategy to go about this is to just ignore waiting for the writes to finish. Right now we are using the sequential queue, and the queue gets blocked by waiting for an action to complete its write to the database.
The important thing to note here is, that writing to the database isn't a heavy operation. On both native and web we just pass our data to the database provider and from there on it's handled in "the background".
However, we are awaiting the finishing of the promises of these write operations, and block the rest of our UI (onyx) operations.

I was considering the ramifications of if we ignored waiting for writes to finish before unblocking the queue.

**Comment by marcaaron** [comment]
Ok, so... is everyone saying they _agree_ that we don't actually need to wait for the writes to happen? 😄
